cassiopeia/solver/HybridSolver.py

```diff
@@ -171,16 +178,18 @@ def solve(
 
             self.__tree = nx.compose(self.__tree, subproblem_tree)
 
-        cassiopeia_tree.populate_tree(self.__tree, layer=layer)
+        # append sample names to the solution and populate the tree
+        samples_tree = self.__append_sample_names(self.__tree, character_matrix)
+        cassiopeia_tree.populate_tree(samples_tree, layer=layer)
+        cassiopeia_tree.collapse_unifurcations()
 
-        # Collapse 0-mutation edges and append duplicate samples
-        cassiopeia_tree.collapse_mutationless_edges(
-            infer_ancestral_characters=True
-        )
-        samples_tree = self.__append_sample_names(
-            cassiopeia_tree.get_tree_topology(), character_matrix
-        )
-        cassiopeia_tree.populate_tree(samples_tree)
+        # collapse mutationless edges
+        if collapse_mutationless_edges:
+            cassiopeia_tree.collapse_mutationless_edges(
+                infer_ancestral_characters=True
+            )
+
+        self.__tree = nx.DiGraph()
```

Comment:
I know I wrote this, but I wonder if there is a more elegant way to do to this than have a hidden variable `__tree` in the object.

I have a solution that works by creating a temporary tree in the `solve` method & passing it to the top-solver; I will push it to this PR.

Reply:
Thanks Matt!